streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,301 @@
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
+
+import pandas as pd
+from pyspark import TaskContext
+from pyspark.sql import SparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.types import (ArrayType, BinaryType, BooleanType, ByteType, DateType,
+                               DayTimeIntervalType, DecimalType, DoubleType, FloatType,
+                               IntegerType, LongType, MapType, ShortType, StringType, StructField,
+                               StructType, TimestampNTZType, TimestampType)
+
+from streaming import MDSWriter
+from streaming.base.format.index import get_index_basename
+from streaming.base.format.mds.encodings import _encodings
+from streaming.base.storage.upload import CloudUploader
+
+logger = logging.getLogger(__name__)
+
+
+def is_iterable(obj: Any) -> bool:
+    """Check if obj is iterable."""
+    return issubclass(type(obj), Iterable)
+
+
+def infer_dataframe_schema(dataframe: DataFrame,
+                           user_defined_cols: Optional[Dict[str, Any]] = None) -> Optional[Dict]:
+    """Retrieve schema to construct a dictionary or do sanity check for MDSWriter.
+
+    Args:
+        dataframe (spark dataframe): dataframe to inspect schema
+        user_defined_cols (dict): user specified schema for MDSWriter
+    Return:
+        If user_defined_cols is None, return schema_dict (dict): column name and dtypes that supported by MDSWriter
+        Else, return None
+    Exceptions:
+        Any of the datatype found to be unsupported by MDSWriter, then raise ValueError
+    """
+    dtype_mapping = {
+        ByteType: 'bytes',
+        ShortType: 'uint64',
+        IntegerType: 'int',
+        LongType: 'int64',
+        FloatType: 'float32',
+        DoubleType: 'float64',
+        DecimalType: 'str_decimal',
+        StringType: 'str',
+        BinaryType: None,
+        BooleanType: None,
+        TimestampType: None,
+        TimestampNTZType: None,
+        DateType: None,
+        DayTimeIntervalType: None,
+        ArrayType: None,
+        MapType: None,
+        StructType: None,
+        StructField: None
+    }
+
+    def map_spark_dtype(spark_data_type: Any):
+        for sparkDtype, mdsDtype in dtype_mapping.items():
+            if isinstance(spark_data_type, sparkDtype) and mdsDtype is not None:
+                return mdsDtype
+        raise ValueError(f'{spark_data_type} is not supported by MDSwriter')
+
+    if user_defined_cols is not None:  # user has provided schema, we just check if mds supports the dtype
+        mds_supported_dtypes = list(dtype_mapping.values())
+        for col_name, mdsDtype in user_defined_cols.items():
+            if col_name not in dataframe.columns:
+                raise ValueError(
+                    f'{col_name} is not a column of input dataframe: {dataframe.columns}')
+            if mdsDtype not in mds_supported_dtypes:
```

Comment:
You should check that it actually matches the dtype schema too, not just hat its' supported

Reply:
It can be a one to many mapping. not possible to forsee all combinations. e.g., user can choose to encode id column with either int64, int32, or str.